flink-table/flink-table-planner/src/main/java/org/apache/calcite/sql/type/SqlTypeAssignmentRule.java

```diff
@@ -0,0 +1,236 @@
+        rule.add(SqlTypeName.DECIMAL);
+        rule.add(SqlTypeName.FLOAT);
+        rule.add(SqlTypeName.REAL);
+        rule.add(SqlTypeName.DOUBLE);
+        rules.add(SqlTypeName.DOUBLE, rule);
+
+        // DECIMAL is assignable from...
+        rule.clear();
+        rule.add(SqlTypeName.TINYINT);
+        rule.add(SqlTypeName.SMALLINT);
+        rule.add(SqlTypeName.INTEGER);
+        rule.add(SqlTypeName.BIGINT);
+        rule.add(SqlTypeName.REAL);
+        rule.add(SqlTypeName.DOUBLE);
+        rule.add(SqlTypeName.DECIMAL);
+        rules.add(SqlTypeName.DECIMAL, rule);
+
+        // VARBINARY is assignable from...
+        rule.clear();
+        rule.add(SqlTypeName.VARBINARY);
+        rule.add(SqlTypeName.BINARY);
+        rules.add(SqlTypeName.VARBINARY, rule);
+
+        // CHAR is assignable from...
+        rules.add(SqlTypeName.CHAR, EnumSet.of(SqlTypeName.CHAR));
+
+        // VARCHAR is assignable from...
+        rule.clear();
+        rule.add(SqlTypeName.CHAR);
+        rule.add(SqlTypeName.VARCHAR);
+        rules.add(SqlTypeName.VARCHAR, rule);
+
+        // BOOLEAN is assignable from...
+        rules.add(SqlTypeName.BOOLEAN, EnumSet.of(SqlTypeName.BOOLEAN));
+
+        // BINARY is assignable from...
+        rule.clear();
+        rule.add(SqlTypeName.BINARY);
+        rule.add(SqlTypeName.VARBINARY);
+        rules.add(SqlTypeName.BINARY, rule);
+
+        // DATE is assignable from...
+        rule.clear();
+        rule.add(SqlTypeName.DATE);
+        rules.add(SqlTypeName.DATE, rule);
+
+        // TIME is assignable from...
+        rule.clear();
+        rule.add(SqlTypeName.TIME);
+        rules.add(SqlTypeName.TIME, rule);
+
+        // TIME WITH LOCAL TIME ZONE is assignable from...
+        rules.add(
+                SqlTypeName.TIME_WITH_LOCAL_TIME_ZONE,
+                EnumSet.of(SqlTypeName.TIME_WITH_LOCAL_TIME_ZONE));
+
+        // TIMESTAMP is assignable from ...
+        rules.add(SqlTypeName.TIMESTAMP, EnumSet.of(SqlTypeName.TIMESTAMP));
+
+        // TIMESTAMP WITH LOCAL TIME ZONE is assignable from...
+        rules.add(
+                SqlTypeName.TIMESTAMP_WITH_LOCAL_TIME_ZONE,
+                EnumSet.of(SqlTypeName.TIMESTAMP_WITH_LOCAL_TIME_ZONE));
+
+        // GEOMETRY is assignable from ...
+        rule.clear();
+        rule.add(SqlTypeName.GEOMETRY);
+        rule.add(SqlTypeName.CHAR);
+        rule.add(SqlTypeName.VARCHAR);
+        rules.add(SqlTypeName.GEOMETRY, rule);
+
+        // ARRAY is assignable from ...
+        rules.add(SqlTypeName.ARRAY, EnumSet.of(SqlTypeName.ARRAY));
+
+        // MAP is assignable from ...
+        rules.add(SqlTypeName.MAP, EnumSet.of(SqlTypeName.MAP));
+
+        // SYMBOL is assignable from ...
+        rules.add(SqlTypeName.SYMBOL, EnumSet.of(SqlTypeName.SYMBOL));
+
+        // ----- FLINK MODIFICATION BEGIN -----
+        // COLUMN_LIST is assignable from ...
+        rules.add(SqlTypeName.COLUMN_LIST, EnumSet.of(SqlTypeName.COLUMN_LIST));
```

Comment:
@snuyanzin could you contribute this line to Calcite? Should be an easy PR if not just a hotfix commit.